### PR TITLE
feat: Go: Add signature key templates WithoutPrefix

### LINKED ISSUE #325: Go: bad signature verification from keyset.Handle created from raw []byte public key
When signing a message with a normally created keyset.Handle from Tink, a prefix is appended to the signature.

If the public key is exchanged with a receiver in raw []bytes (ie marshalled elliptic curve bytes) and the receiver tries to create keyset.Handle from these raw bytes (by pre-building the associated protobuf for the given curve used in the signature) then verifies the signature, it fails since raw type does not have a prefix.

Particularly, this code snippet for raw keys fails (from Verify() in `go/signature/verifier_factory.go`):
```
	// try raw keys
	entries, err = v.ps.RawEntries()
	if err == nil {
		for i := 0; i < len(entries); i++ {
			var verifier = (entries[i].Primitive).(tink.Verifier)
			if err = verifier.Verify(signature, data); err == nil {
				return nil
			}
		}
	}
```

It fails because signing was done with valid Tink created keyset.Handle which has Key.OutputPrefixType = `tinkpb.OutputPrefixType_TINK`

But for verification, as the recipient has only the []bytes representing the raw marshalled EC key value, tinkpb.Keyset_key is built with OutPrefixType = `tinkpb.OutputPrefixType_RAW` and KeyId = 1 (the recipient does not know of KeyId, we just default it to 1 for the purpose of key verification).

A possible fix I can think of this is to run Verify() with the prefix trimmed (for raw keys only) and return nil if successful (I have tried it locally and it does return a successful verification). Here is the updated code snippet for raw keys for this fix:
```
	// try raw keys
	entries, err = v.ps.RawEntries()
	if err == nil {
		for i := 0; i < len(entries); i++ {
			var verifier = (entries[i].Primitive).(tink.Verifier)
			if err = verifier.Verify(signature, data); err == nil {
				return nil
			}

			if len(signature) > cryptofmt.NonRawPrefixSize {
				// try signature by stripping NonRaw Prefix
				strippedSig := signature[cryptofmt.NonRawPrefixSize:]
				if err = verifier.Verify(strippedSig, data); err == nil {
					return nil
				}
			}
		}
	}
```

### PR DESCRIPTION
**Pull request by Baha-sk**
To support using imported keys from raw key bytes, Tink
users must use signing key templates without prefix to
both sign and verify signatures.

This patch also fixes some `go import` formatting of some
files.

closes #325

Signed-off-by: Baha Shaaban <baha.shaaban@securekey.com>

### CODE REVIEW

File: go/signature/signature_key_templates_test.go [hunk 1]
```diff
@@ -19,65 +19,141 @@ import (
...
-	encoding ecdsapb.EcdsaSignatureEncoding) error {
-	if template.TypeUrl != testutil.ECDSASignerTypeURL {
-		return fmt.Errorf("incorrect typeurl: expect %s, got %s", testutil.ECDSASignerTypeURL, template.TypeUrl)
+	encoding ecdsapb.EcdsaSignatureEncoding,
+	prefixType tinkpb.OutputPrefixType) error {
+	if template.TypeUrl != typeURL {
+		return fmt.Errorf("incorrect typeurl: expect %s, got %s", typeURL, template.TypeUrl)
+	}
+
+	if template.OutputPrefixType.String() != prefixType.String() {
```
Comment (thaidn) [inline go/signature/signature_key_templates_test.go]: I think we can compare the prefix type directly, i.e.,

if template.OutputPrefixType != prefixType {
   ...
}

File: go/signature/signature_key_templates_test.go [hunk 2]
```diff
@@ -19,65 +19,141 @@ import (
...
+			typeURL:     testutil.ED25519SignerTypeURL,
+			sigTemplate: signature.ED25519KeyTemplate(),
+			prefixType:  tinkpb.OutputPrefixType_TINK,
+		},
+		{
+			tcName:      "test ED25519 with RAW output prefix",
+			typeURL:     testutil.ED25519SignerTypeURL,
+			sigTemplate: signature.ED25519KeyWithoutPrefixTemplate(),
+			prefixType:  tinkpb.OutputPrefixType_RAW,
+		},
```
Comment (thaidn) [inline go/signature/signature_key_templates_test.go]: Please add a new test for ED25519 key templates. They aren't different from ECDSA key templates.

File: go/signature/signature_key_templates_test.go [hunk 3]
```diff
@@ -19,65 +19,141 @@ import (
...
+	prefixType tinkpb.OutputPrefixType) error {
+	if template.TypeUrl != typeURL {
+		return fmt.Errorf("incorrect typeurl: expect %s, got %s", typeURL, template.TypeUrl)
+	}
+
+	if template.OutputPrefixType.String() != prefixType.String() {
+		return fmt.Errorf("incorrect outputPrefixType: expect: %v, got %v", prefixType, template.OutputPrefixType)
 	}
+
+	if typeURL == testutil.ED25519SignerTypeURL {
```
Comment (thaidn) [inline go/signature/signature_key_templates_test.go]: Please remove this.

File: go/integration/hcvault/go.mod [hunk 4]
```diff
@@ -3,7 +3,7 @@ module github.com/google/tink/go/integration/hcvault
 go 1.12
```
Comment (thaidn) [inline go/integration/hcvault/go.mod]: Please revert this change, as it's out of scope. You're welcome to send another PR for this.

File: go/signature/signature_key_templates_test.go (hunk 1 above)
Comment (Baha-sk) [reply go/signature/signature_key_templates_test.go]: Indeed, because  the underlying type of OutputPrefixType is uint32.

Done.

File: go/integration/hcvault/go.mod (hunk 4 above)
Comment (Baha-sk) [reply go/integration/hcvault/go.mod]: unfortunately my IDE keeps updating this automatically .. I will try to revet this outside my IDE

File: go/signature/signature_key_templates_test.go (hunk 2 above)
Comment (Baha-sk) [reply go/signature/signature_key_templates_test.go]: The two elements above are for ED25519 tests. I'm not sure I understand the request.

File: go/signature/signature_key_templates_test.go (hunk 3 above)
Comment (Baha-sk) [reply go/signature/signature_key_templates_test.go]: ED25519 templates don't have key format, so the following tests are unnecessary:

https://github.com/google/tink/blob/master/go/signature/signature_key_templates.go#L75

I think you want me to split Ed25519 key verification from ECDSA key verification in two separate functions?

File: go/integration/hcvault/go.mod (hunk 4 above)
Comment (Baha-sk) [reply go/integration/hcvault/go.mod]: changes reverted

File: go/signature/signature_key_templates_test.go (hunk 2 above)
Comment (Baha-sk) [reply go/signature/signature_key_templates_test.go]: I see what you mean, I need to create another Test function for EDDSA, similar to ECDSA.

I will split the functions in a bit.

Comment (Baha-sk) [reply go/signature/signature_key_templates_test.go]: created TestEDDSAKeyTemplates() for ED25519 key templates

File: go/signature/signature_key_templates_test.go (hunk 3 above)
Comment (Baha-sk) [reply go/signature/signature_key_templates_test.go]: I split ECDSA key templates (TestECDSAKeyTemplates) tests from EDDSA key templates (TestEDDSAKeyTemplates).

ECDSA calls `checkECDSAKeyTemplate` to check for common key template fields (by calling `checkDSAKeyTemplate`) and has additional fields check (namely key format params).

EDDSA calls `checkDSAKeyTemplate` for common fields only.

**Comment by Baha-sk** [comment]
Please let me know if there are any pending changes required for this PR to be merged.

File: go/signature/signature_key_templates_test.go [hunk 5]
```diff
@@ -19,65 +19,170 @@ import (
...
+			tcName:      "ED25519 with RAW output prefix",
+			typeURL:     testutil.ED25519SignerTypeURL,
+			sigTemplate: signature.ED25519KeyWithoutPrefixTemplate(),
+			prefixType:  tinkpb.OutputPrefixType_RAW,
+		},
+	}
+
+	for _, tt := range flagTests {
+		t.Run("Test EDDSA - "+tt.tcName, func(t *testing.T) {
+			err := checkDSAKeyTemplate(tt.sigTemplate,
```
Comment (thaidn) [inline go/signature/signature_key_templates_test.go]: DSA is actually a name of another digital signature algorithm, so this is while confusing.

I suggest renaming is to checkKeyTypeAndOutputPrefix

File: go/signature/verifier_factory.go [hunk 6]
```diff
@@ -22,8 +22,8 @@ import (
 	"github.com/google/tink/go/core/primitiveset"
 	"github.com/google/tink/go/core/registry"
 	"github.com/google/tink/go/keyset"
-	"github.com/google/tink/go/tink"
 	tinkpb "github.com/google/tink/go/proto/tink_go_proto"
```
Comment (thaidn) [inline go/signature/verifier_factory.go]: could you also revert this change? it'll interfere with our merging process.

File: go/subtle/signature/ed25519_signer.go [hunk 7]
```diff
@@ -15,7 +15,7 @@
 package signature
 
 import (
-	"golang.org/x/crypto/ed25519"
```
Comment (thaidn) [inline go/subtle/signature/ed25519_signer.go]: this change is okay to keep.

File: go/signature/signature_key_templates.go [hunk 8]
```diff
@@ -67,12 +113,22 @@ func createECDSAKeyTemplate(hashType commonpb.HashType, curve commonpb.EllipticC
...
 	}
 }
+
+// ED25519KeyWithoutPrefixTemplate is a KeyTemplate that generates a new ED25519 private key.
+func ED25519KeyWithoutPrefixTemplate() *tinkpb.KeyTemplate {
+	return &tinkpb.KeyTemplate{
+		TypeUrl: ed25519SignerTypeURL,
+		OutputPrefixType: tinkpb.OutputPrefixType_RAW,
+	}
+}
```
Comment (thaidn) [inline go/signature/signature_key_templates.go]: Please add a newline at the end of this file.

File: go/signature/verifier_factory.go (hunk 6 above)
Comment (Baha-sk) [reply go/signature/verifier_factory.go]: weird, but this is done by normal `go fmt` .. I'll revert

File: go/signature/signature_key_templates.go (hunk 8 above)
Comment (Baha-sk) [reply go/signature/signature_key_templates.go]: forgot to run `go fmt` on this file.. good catch.. I'll update

File: go/subtle/signature/ed25519_signer.go (hunk 7 above)
Comment (Baha-sk) [reply go/subtle/signature/ed25519_signer.go]: will revert

File: go/signature/signature_key_templates_test.go (hunk 5 above)
Comment (Baha-sk) [reply go/signature/signature_key_templates_test.go]: done

File: go/signature/verifier_factory.go (hunk 6 above)
Comment (Baha-sk) [reply go/signature/verifier_factory.go]: reverted change

File: go/subtle/signature/ed25519_signer.go (hunk 7 above)
Comment (Baha-sk) [reply go/subtle/signature/ed25519_signer.go]: reverted

File: go/signature/signature_key_templates.go (hunk 8 above)
Comment (Baha-sk) [reply go/signature/signature_key_templates.go]: done

File: go/signature/signature_key_templates_test.go [hunk 9]
```diff
@@ -19,65 +19,170 @@ import (
 	"testing"
 
 	"github.com/golang/protobuf/proto"
-	"github.com/google/tink/go/signature"
-	"github.com/google/tink/go/testutil"
 	commonpb "github.com/google/tink/go/proto/common_go_proto"
 	ecdsapb "github.com/google/tink/go/proto/ecdsa_go_proto"
 	tinkpb "github.com/google/tink/go/proto/tink_go_proto"
+	"github.com/google/tink/go/signature"
```
Comment (Baha-sk) [inline go/signature/signature_key_templates_test.go]: I guess these should go back before  the proto imports? I will revert the import changes..

Comment (Baha-sk) [reply go/signature/signature_key_templates_test.go]: revert the order of these imports as well

File: go/signature/signature_key_templates_test.go (hunk 2 above)
Comment (Baha-sk) [reply go/signature/signature_key_templates_test.go]: Renamed function to `TestED25519KeyTemplates()` instead.

**Comment by Baha-sk** [comment]
I just rebased this PR.. there's no additional changes
